koku/api/report/view.py

```diff
@@ -27,11 +27,85 @@
+from api.tags.ocp.ocp_tag_query_handler import OCPTagQueryHandler
+from api.tags.serializers import AWSTagsQueryParamSerializer, OCPTagsQueryParamSerializer
 from api.utils import UnitConverter
+from reporting.provider.aws.models import AWSTagsSummary
+from reporting.provider.ocp.models import OCPUsagePodLabelSummary
+
 
 LOG = logging.getLogger(__name__)
 
 
+class ClassMapper(object):
+    """Data structure object to organize class references."""
+
+    CLASS_MAP = [{'provider': 'aws', 'reports': [{'report': 'default',
+                                                  'serializer': QueryParamSerializer,
+                                                  'query_handler': AWSReportQueryHandler,
+                                                  'tag_handler': AWSTagsSummary},
+                                                 {'report': 'tags',
+                                                  'serializer': AWSTagsQueryParamSerializer,
+                                                  'query_handler': AWSTagQueryHandler,
+                                                  'tag_handler': AWSTagsSummary}]},
+                 {'provider': 'ocp', 'reports': [{'report': 'default',
+                                                  'serializer': OCPInventoryQueryParamSerializer,
+                                                  'query_handler': OCPReportQueryHandler,
+                                                  'tag_handler': OCPUsagePodLabelSummary},
+                                                 {'report': 'charge',
+                                                  'serializer': OCPChargeQueryParamSerializer,
+                                                  'query_handler': OCPReportQueryHandler,
+                                                  'tag_handler': OCPUsagePodLabelSummary},
+                                                 {'report': 'tags',
+                                                  'serializer': OCPTagsQueryParamSerializer,
+                                                  'query_handler': OCPTagQueryHandler,
+                                                  'tag_handler': OCPUsagePodLabelSummary}]}]
+
+    def reports(self, provider):
+        """Return list of report dictionaries."""
+        for item in self.CLASS_MAP:
+            if provider == item.get('provider'):
+                return item.get('reports')
+
+    def report_types(self, provider):
+        """Return list of report names."""
+        reports = self.reports(provider)
+        return [rep.get('report') for rep in reports]
+
+    def get_report(self, provider, report):
+        """Return the specified report dict.
+
+        Return default report dict if named report not found.
+
+        """
+        reports = self.reports(provider)
+        for rep in reports:
+            if report == rep.get('report'):
+                return rep
+        return self.get_report(provider, 'default')
```

Comment:
The default here could get get us into trouble if we forget to implement a new entry in the map or misspell something. It will happily supply things that we might not intend. Since we have to explicitly define each API endpoint, I don't think it's a bad idea to be explicit with the report types here.

Reply:
The problem I was solving was just that most reports end up using the same serializer and query handler. So, I opted for a defaulting mechanism instead of having a hugely redundant data structure.

I can reformulate it, but I agree with you that trade-off exists and I'm okay if you'd prefer to pick having an explicit, but highly redundant data structure instead. :slightly_smiling_face:

Reply:
Nah, I suppose it will fail in the query handler if the report type is not implemented there.